components/eam/src/physics/cam/zm_microphysics.F90

```diff
@@ -0,0 +1,3337 @@
+real(r8) :: rin       ! radius of contact nuclei
+real(r8) :: pi        ! pi
+real(r8) :: mmult
+
+! for Bergeron process (Rotstayn et al.2000)
+real(r8) :: Ka_b       ! thermal conductivity of air(J/m/s/K)
+real(r8) :: Ls_b       ! latent heat of sublimation of water(J/kg)
+real(r8) :: Rv_b       ! specigic gas constant for water vapour(J/kg/K)
+real(r8) :: alfa_b     ! parameter for ice crystal habit
+real(r8) :: rhoi13     ! rhoi**(1/3)
+real(r8) :: c23        ! 2/3  
+
+real(r8) ::  cons14,cons16,cons17,cons18,cons19,cons24,cons25, cons31, cons32, cons41
+
+real(r8) :: droplet_mass_25um
+
+! contact freezing due to dust
+! dust number mean radius (m), Zender et al JGR 2003 assuming number mode radius of 0.6 micron, sigma=2
+real(r8), parameter :: rn_dst1 = 0.258e-6_r8
+real(r8), parameter :: rn_dst2 = 0.717e-6_r8
+real(r8), parameter :: rn_dst3 = 1.576e-6_r8
+real(r8), parameter :: rn_dst4 = 3.026e-6_r8
+
+! smallest mixing ratio considered in microphysics
+real(r8), parameter :: qsmall = 1.e-18_r8
+
+
+type, public :: ptr2d
+   real(r8), pointer :: val(:,:)
+end type ptr2d
+
+! Aerosols
+type :: zm_aero_t
+
+   ! Aerosol treatment
+   character(len=5) :: scheme  ! either 'bulk' or 'modal'
+
+   ! Bulk aerosols
+   integer :: nbulk    =  0 ! number of bulk aerosols affecting climate
+   integer :: idxsul   = -1 ! index in aerosol list for sulfate
+   integer :: idxdst1  = -1 ! index in aerosol list for dust1
+   integer :: idxdst2  = -1 ! index in aerosol list for dust2
+   integer :: idxdst3  = -1 ! index in aerosol list for dust3
+   integer :: idxdst4  = -1 ! index in aerosol list for dust4
+   integer :: idxbcphi = -1 ! index in aerosol list for Soot (BCPHI)
+
+   real(r8),    allocatable :: num_to_mass_aer(:)  ! conversion of mmr to number conc for bulk aerosols
+   type(ptr2d), allocatable :: mmr_bulk(:)         ! array of pointers to bulk aerosol mmr
+   real(r8),    allocatable :: mmrg_bulk(:,:,:)    ! gathered bulk aerosol mmr
+
+   ! Modal aerosols
+   integer                  :: nmodes = 0      ! number of modes
+   integer,     allocatable :: nspec(:)        ! number of species in each mode
+   type(ptr2d), allocatable :: num_a(:)        ! number mixing ratio of modes (interstitial phase)
+   type(ptr2d), allocatable :: mmr_a(:,:)      ! species mmr in each mode (interstitial phase)
+   real(r8),    allocatable :: numg_a(:,:,:)   ! gathered number mixing ratio of modes (interstitial phase)
+   real(r8),    allocatable :: mmrg_a(:,:,:,:) ! gathered species mmr in each mode (interstitial phase)
+   real(r8),    allocatable :: voltonumblo(:)  ! volume to number conversion (lower bound) for each mode
+   real(r8),    allocatable :: voltonumbhi(:)  ! volume to number conversion (upper bound) for each mode
+   real(r8),    allocatable :: specdens(:,:)   ! density of modal species
+   real(r8),    allocatable :: spechygro(:,:)  ! hygroscopicity of modal species
+
+   integer :: mode_accum_idx  = -1  ! index of accumulation mode
```

Comment:
@mingxuanwupnnl , please help check here if the hardwired code here exclude other aerosol options.

Reply:
@hwangacme yes, I think there are two issues for these indexes, especially for coarse_dust_idx and coarse_nacl_idx. They are related to lines (2192-2222) for ice nucleations.